torch/utils/data/_utils/index_utils.py

```diff
@@ -0,0 +1,146 @@
+from math import isqrt
+
+
+# from random import Random
+
+
+class Range:
+    size: int
+
+    # for slicing
+    start: int
+    stop: int
+    step: int
```

Comment:
Besides, size seems duplicated with start/stop/step. There is a chance of conflicts between size and the other three parameters like your `_slice` function.

Reply:
> Hey, @ejguan , got a question here. I went though the code earlier yesterday, it seems the prime probe happened when Permutation get created or _slice() is called (to get 1st prime >= size). The time complexity is high (O log(p)p), but I think normally what we do is to create a random permutation, and slice it into multiple subsets if necessary, and after that, the heavy prime probe calculation won't happen anymore. The number of subsets we slice is where this might get much slower, I am not sure how many subsets people normally creates, if this is not a very big number (I think million level is still fine), it should not affect the overall performance that much. I might missed some important use cases here, correct me if I am wrong.

Slicing does not cause a prime to be searched. Only initial creation. That's also why there's both a size and a start/stop/step.

Also there's much faster primehood algorithms than the brute-force method I use, but I couldn't find a readily-available implementation and didn't want to try to copy&paste one from the net.

Reply:
What conflicts?

Reply:
You are using size to calculate the abs_index, but you never verify size is correct.

Reply:
Just gonna add this here for other reviewers that we discussed size in the comments. Specifically, that the intent is for size to never change, that Range features wrap-around behavior and size is distinct from length, which is calculated based on start, stop, and step. Size should equal `len(dataset)`. It defines the indices which can be returned. (Perhaps calling it Range and calling constructor keyword arguments start/stop/step might cause confusion. The semantics are different compared to regular `range`.)